src/cpp/cpp_typecheck_code.cpp

```diff
@@ -55,22 +55,29 @@ void cpp_typecheckt::typecheck_code(codet &code)
     // as an extension, we support indexed access into signed/unsigned
     // bitvectors, typically used with __CPROVER::(un)signedbv<N>
     exprt &expr = code.op0();
-    if(
-      expr.operands().size() == 2 && expr.op0().id() == ID_index &&
-      expr.op0().operands().size() == 2)
+
+    if(expr.operands().size() == 2)
     {
-      exprt array = expr.op0().op0();
-      typecheck_expr(array);
+      auto &binary_expr = to_binary_expr(expr);
 
-      if(array.type().id() == ID_signedbv || array.type().id() == ID_unsignedbv)
+      if(binary_expr.op0().id() == ID_index)
       {
-        shl_exprt shl{from_integer(1, array.type()), expr.op0().op1()};
-        exprt rhs =
-          if_exprt{equal_exprt{expr.op1(), from_integer(0, array.type())},
-                   bitand_exprt{array, bitnot_exprt{shl}},
-                   bitor_exprt{array, shl}};
-        expr.op0() = expr.op0().op0();
-        expr.op1() = rhs;
+        exprt array = to_index_expr(binary_expr.op0()).array();
```

Comment:
I'd suggest to introduce a temporary `index_exprt` as `binary_expr.op0()` is used several times below.

Reply:
I've avoided that, since that gets destroyed in line 78.